### ISSUE TITLE
torch.softmax gives wierd result under specific condition on cuda double tensors with multiple rows and 781 columns.

### ISSUE THREAD

**Issue by Xenadon** [pytorch/pytorch#147284]
### 🐛 Describe the bug

`torch.softmax(torch.zeros(c,781, dtype=torch.float64, device='cuda'),dim=1)` gives wrong answer for `c > 1`.

```python
import torch
print(torch.softmax(torch.zeros(1,781, dtype=torch.float64, device='cuda'),dim=1)) # correct
print(torch.softmax(torch.zeros(2,781, dtype=torch.float64, device='cuda'),dim=1)) # wrong
print(torch.softmax(torch.zeros(3,781, dtype=torch.float64, device='cuda'),dim=1)) # wrong
print(torch.softmax(torch.zeros(4,781, dtype=torch.float64, device='cuda'),dim=1)) # wrong
print(torch.softmax(torch.zeros(5,781, dtype=torch.float64, device='cuda'),dim=1)) # wrong
print(torch.softmax(torch.zeros(2,780, dtype=torch.float64, device='cuda'),dim=1)) # correct
print(torch.softmax(torch.zeros(2,782, dtype=torch.float64, device='cuda'),dim=1)) # correct
```

output:
```
tensor([[0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013,
	 # ... (many lines of 0.0013)
         0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013,
         0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)

tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006]],
       device='cuda:0', dtype=torch.float64)

tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)

tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006]],
       device='cuda:0', dtype=torch.float64)

tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)

tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013],
        [0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)

tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013],
        [0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)

```

### Versions

```
Collecting environment information...
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.12.8 | packaged by Anaconda, Inc. | (main, Dec 11 2024, 16:31:09) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.82
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 556.13
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i9-14900HX
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             1
BogoMIPS:                             4838.39
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            768 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             32 MiB (16 instances)
L3 cache:                             36 MiB (1 instance)
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] equitorch==0.1
[pip3] numpy==2.2.2
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.5.1
[pip3] torch_cluster==1.6.3+pt25cu124
[pip3] torch-geometric==2.6.1
[pip3] torch_scatter==2.1.2+pt25cu124
[pip3] torch_sparse==0.6.18+pt25cu124
[pip3] torch_spline_conv==1.2.2+pt25cu124
[pip3] triton==3.2.0
[conda] equitorch                 0.1                      pypi_0    pypi
[conda] numpy                     2.2.2                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] torch                     2.5.1                    pypi_0    pypi
[conda] torch-cluster             1.6.3+pt25cu124          pypi_0    pypi
[conda] torch-geometric           2.6.1                    pypi_0    pypi
[conda] torch-scatter             2.1.2+pt25cu124          pypi_0    pypi
[conda] torch-sparse              0.6.18+pt25cu124          pypi_0    pypi
[conda] torch-spline-conv         1.2.2+pt25cu124          pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi
```

**Comment by manncodes** [comment]
Interesting. 

Some workarounds that worked:

Use float32 instead of float64:
```python
tensor = torch.zeros(c, 781, dtype=torch.float32, device='cuda')
```

2. Process on CPU:
```python
result = torch.softmax(tensor.cpu(), dim=1).cuda()
```

3. Pad to 782 columns:
```python
tensor = torch.zeros(c, 782, dtype=torch.float64, device='cuda')
result = torch.softmax(tensor, dim=1)[:, :781]
```


I think it's very likely to be a  CUDA-specific implementation issue with softmax for float64 tensors. Looking at this file:
```
aten/src/ATen/native/cuda/SoftMax.cu
```

**Comment by mikaylagawarecki** [comment]
I can't repro this


```python
>>> import torch
>>> print(torch.softmax(torch.zeros(2,781, dtype=torch.float64, device='cuda'),dim=1)) 
tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013],
        [0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)
>>> 
```

**Comment by tringwald** [comment]
I can reproduce this on `torch==2.6.0+cu124` with an RTX 4090.

```
tensor([[0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013,
         0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013,
         [...]
         0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013,
         0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)
tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006]],
       device='cuda:0', dtype=torch.float64)
tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)
tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006]],
       device='cuda:0', dtype=torch.float64)
tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0013, 0.0013, 0.0013,  ..., 0.0006, 0.0006, 0.0006],
        [0.0006, 0.0006, 0.0006,  ..., 0.0006, 0.0006, 0.0006],
        [0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)
tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013],
        [0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)
tensor([[0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013],
        [0.0013, 0.0013, 0.0013,  ..., 0.0013, 0.0013, 0.0013]],
       device='cuda:0', dtype=torch.float64)
```

**Comment by tringwald** [comment]
I can't reproduce it on latest `main` though. I think this was fixed by https://github.com/pytorch/pytorch/issues/143644.

**Comment by u7k4rs6** [comment]
Independent reproduction, plus a characterization of exactly which `dim_size` values are affected. Posting here rather than opening a new issue since this looks like the same defect.

**Environment:** torch 2.6.0+cu124, CUDA 12.4, driver 595.84, RTX 4060 Laptop (sm_89), Linux.

781 reproduces here, and so do 239 other widths. Since @mikaylagawarecki could not reproduce and @tringwald could on a 4090, an exact predicate may help narrow what distinguishes the machines.

### The failing set is exactly characterizable

Sweeping every width from 2 to 4096 with `torch.zeros(2, w, dtype=torch.float64, device="cuda")` and testing whether the probabilities sum to 1, 240 widths fail. All 240 are predicted with no false positives or negatives by:

```
512 < dim_size < 1024   and   dim_size is odd   and   dim_size % 32 != 31
```

with `rows >= 2` (a single row is always correct). 781 satisfies this: it is odd, in range, and `781 % 32 == 13`. So does 513, which is the smallest. Nothing at or above 1024 fails, and nothing at or below 512 does.

```python
import torch

observed = set()
for w in range(2, 4097):
    s = torch.softmax(torch.zeros(2, w, dtype=torch.float64, device="cuda"), dim=1).sum(-1)
    if float((s - 1.0).abs().max()) > 1e-12:
        observed.add(w)

predicted = {w for w in range(2, 4097)
             if 512 < w < 1024 and w % 2 == 1 and w % 32 != 31}

print(len(observed), observed == predicted)   # 240 True
```

The `% 32 != 31` exception is the part I find most suggestive: widths 543, 575, 607, ... 1023 are all odd and in range but are correct. That is consistent with a tail-handling boundary in the vectorized path rather than with anything about the input values.

### Stating it as a sum, which is unambiguous

The output is not merely imprecise, it is not a probability distribution:

```
width 512: max |sum - 1| = 4.441e-16
width 513: max |sum - 1| = 4.998e-01   <-- wrong
width 514: max |sum - 1| = 2.220e-16
width 769: max |sum - 1| = 6.584e-01   <-- wrong
```

### Scope

All at width 513, 8 rows, compared against the CPU result:

| case | max abs error |
|---|---|
| `softmax`, float64, CUDA | 4.99e-01 |
| `log_softmax`, float64, CUDA | 6.93e-01 |
| `softmax`, float64, CUDA, non-contiguous input | 4.99e-01 |
| `softmax`, float64, CUDA, `use_deterministic_algorithms(True)` | 4.99e-01 |
| `softmax`, float32, CUDA | 3.28e-07 |
| `softmax`, float64, CPU | 0.0 |
| `logsumexp`, float64, CUDA | 0.0 |
| explicit max-subtract softmax, float64, CUDA | 6.66e-16 |

So `log_softmax` is affected as well, deterministic mode does not avoid it, and `logsumexp` is not affected. float32 and CPU are clean.

Only float64 is affected, only for `512 < dim_size < 1024`, which is the range where `next_power_of_two(dim_size)` is 1024. At 8 bytes per element that is the largest per-thread working set the persistent warp path handles, so `aten/src/ATen/native/cuda/SoftMax.cu` and its ILP or vectorized tail handling seems the likely area, though I have not confirmed that by reading the generated code.

A standalone script that prints all of the above is [here](https://github.com/u7k4rs6/LockStep/blob/main/scripts/repro_torch_softmax_fp64.py).

Happy to run further diagnostics on this machine if a specific hypothesis would help.
